### PR TITLE
Add ability to throttle merges/mutations

### PR DESCRIPTION
**Pull request by azat**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add ability to throttle merges/mutations (`max_mutations_bandwidth_for_server`/`max_merges_bandwidth_for_server`)

Main motivation was to has an ability to throttle background tasks, to
avoid affecting queries.

To new server settings had been added for this:
- max_mutations_bandwidth_for_server
- max_merges_bandwidth_for_server

Note, that they limit only reading, since usually you will not write
more data then you read, but sometimes it is possible in case of ALTER
UPDATE.

But for now, to keep things simple, I decided to limit this with only
2 settings instead of 4.

Note, that if the write throttling will be needed, then they can use the
same settings, and just create new throttler for write.

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 738f20e6ac217f672013fecbbcd84a42cfa969bf with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/57877/738f20e6ac217f672013fecbbcd84a42cfa969bf/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/7213395660">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/57877/738f20e6ac217f672013fecbbcd84a42cfa969bf/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Comment by azat** [comment]
>Stateless tests (debug, s3 storage) [5/6] — fail: 2, passed: 986, skipped: 23

- https://github.com/ClickHouse/ClickHouse/pull/58012

**Comment by azat** [comment]
Can someone take a look please? (In the mean time I will fix conflicts)

**Comment by azat** [comment]
>Upgrade check (msan) — check_status.tsv doesn't exists

```
2023.12.25 00:56:02.925970 [ 187827 ] {} <Error> Application: Caught exception while loading metadata: Code: 695. DB::Exception: Load job 'load table test_3.testCleanupR1' failed: Code: 36. DB::Exception: Wrong number of engine arguments.: Cannot attach table `test_3`.`testCleanupR1` from metadata file /var/lib/clickhouse/store/fcf/fcf5a2b4-8900-4a4d-a223-f2a4c28c28d8/testCleanupR1.sql from query ATTACH TABLE test_3.testCleanupR1 UUID 'b24abdb6-5810-462c-9504-f92d80251cb1' (`uid` String, `version` UInt32, `is_deleted` UInt8) ENGINE = ReplicatedReplacingMergeTree('/clickhouse/test_3/tables/test_cleanup/', 'r1', version, is_deleted) ORDER BY uid SETTINGS enable_vertical_merge_algorithm = 1, vertical_merge_algorithm_min_rows_to_activate = 0, vertical_merge_algorithm_min_columns_to_activate = 0, min_rows_for_wide_part = 0, min_bytes_for_wide_part = 0, index_granularity = 41917, ratio_of_defaults_for_sparse_serialization = 1., replace_long_file_name_to_hash = 1, max_file_name_length = 56, merge_max_block_size = 14489, prefer_fetch_merged_part_size_threshold = 10737418240, min_merge_bytes_to_use_direct_io = 1, index_granularity_bytes = 16945590, allow_vertical_merges_from_compact_to_wide_parts = 1, marks_compress_block_size = 88601, primary_key_compress_block_size = 16356. (BAD_ARGUMENTS), Stack trace (when copying this message, always include the lines below):
```

**Comment by azat** [comment]
@serxa maybe you can take a look?

**Comment by serxa** [comment]
Thanks for pointing, I completely missed when this PR was created

File: src/Storages/MergeTree/MergeTreeSequentialSource.cpp [hunk 1]
```diff
@@ -142,10 +146,25 @@ MergeTreeSequentialSource::MergeTreeSequentialSource(
         columns_for_reader = data_part->getColumns().addTypes(columns_to_read);
     }
 
-    ReadSettings read_settings;
+    const auto & context = storage.getContext();
+    ReadSettings read_settings = context->getReadSettings();
+    read_settings.read_from_filesystem_cache_if_exists_otherwise_bypass_cache = true;
+    /// It does not make sense to use pthread_threadpool for background merges/mutations
+    /// And also to preserve backward compatibility
+    read_settings.local_fs_method = LocalFSReadMethod::pread;
```
Comment (serxa) [inline src/Storages/MergeTree/MergeTreeSequentialSource.cpp:154]: This change looks unrelated, let's not mix things?

**Review by serxa — APPROVED** [review]

File: src/Storages/MergeTree/MergeTreeSequentialSource.cpp (hunk 1 above)
Comment (azat) [reply src/Storages/MergeTree/MergeTreeSequentialSource.cpp:154]: This is actually required. Before default ReadSettings was used (which has local_fs_method=pread), now they are used from Context (which has local_fs_method=pread_threadpool). And we do need ReadSettings from context to get throttling.

Comment (serxa) [reply src/Storages/MergeTree/MergeTreeSequentialSource.cpp:154]: Indeed
